### PR TITLE
Add basic components

### PR DESCRIPTION
**Pull request by vivangkumar**
### What?

Making this a PR so that it is easier to review. 
Adds the minimum required features so we can push through with writing the chatkit-server-go SDK. Most of this is copied over from the price elements-client-go library, with the exception of some additions.

Also contains tests.

TODO (after merging)
- Changelog
- Add GoDoc link 

Edit:
Also added some helper functions so that SDK's built on top of this can make use of the instance locator parsing functions without having to rewrite them as it is going to be common to all products.

### Why?

Because we don't have a platform library in Go, and it is not standardised when it comes to other platform SDK's such as Ruby and Node. This is by no means feature complete, but supports all the basic features to support writing product SDK's.

----

- [ ] README updated if you changed the API?

----

CC @pusher/sigsdk

### CODE REVIEW

**Comment by vivangkumar** [comment]
[Write a basic pusher-platform-go library for chatkit-server-go](https://trello.com/c/hWKrocQz/1328-write-a-basic-pusher-platform-go-library-for-chatkit-server-go)

File: instance.go [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+	}, nil
+}
+
+// Request allows making HTTP requests to services
+func (i *instance) Request(ctx context.Context, options RequestOptions) (*http.Response, error) {
+	path := i.scopePath(options.Path)
+
+	var jwt *string
+	if options.Jwt == nil {
+		tokenResponse, err := i.Authenticator().GenerateAccessToken(AuthenticateOptions{Su: true})
```
Comment (BenPope) [inline instance.go]: Should `Su: true` be hard-coded?

**Review by BenPope — COMMENTED** [review]
I'd like to see some test coverage of Authenticator and Instance.

File: instance.go (hunk 1 above)
Comment (vivangkumar) [reply instance.go]: Yeah, if there isn't a JWT, we have to fetch one.

Comment (hamchapman) [reply instance.go]: I'm not sure `Su: true` should be the "fallback" behaviour though. It's powerful enough that I think you should have to provide it as an explicit option or use a specialised function like `SuRequest` or whatever.

Comment (hamchapman) [reply instance.go]: Also, su tokens are (at least for now) a Chatkit-specific concern, so shouldn't _really_ be something that appears at all in a platform SDK. In the Chatkit Go SDK we should probably just have a `SuRequest` that sets up a `Request` call with the appropriate options.

Comment (vivangkumar) [reply instance.go]: Hmmm, I'm quite sure I saw this in a platform library...

EDIT: https://github.com/pusher/pusher-platform-node/blob/master/src/instance.ts#L81

Although you are definitely right..

Comment (vivangkumar) [reply instance.go]: I have now removed it..although it does show that the platform libraries all are written that way which should probably be fixed.

**Comment by vivangkumar** [comment]
@BenPope Tests added! Ready for review again.

**Comment by BenPope** [comment]
Most of the things in common.go are actually only used in auth.

I wonder if it's worth splitting auth, instance, and client into their own subpackages? That would allow helpers to move inside instance and be kept private. I think it will also clean up naming a bit.

There's a typo in the example in the doc; is it worth moving them into part of the go documentation so that they are checked?  The godoc is looking a little sparse.

File: README.md [hunk 2]
```diff
@@ -69,17 +92,22 @@ import (
...
 	ServiceName: "<SERVICE-NAME-TO-CONNECT-TO>",
 	ServiceVersion: "<SERVICE-VERSION>",
 })
+if err != nil {
+	...
+}
 
 ctx := context.background()
+jwt := "your-jwt-token"
 resp, err := instsance.Request(ctx, RequestOptions{
```
Comment (BenPope) [inline README.md]: Typo

File: README.md [hunk 3]
```diff
@@ -34,22 +60,37 @@ Instance objects provide access to the `Authenticator` which can be used to buil
...
+	ServiceName: "<SERVICE-NAME-TO-CONNECT-TO>",
+	ServiceVersion: "<SERVICE-VERSION>",
+})
+if err != nil {
+	...
+}
+
 userID := "abc"
-authResponse, err := instance.Authenticator().Authenticate(platform.AuthenticatePayload{
+authResponse, err := app.Authenticator().Authenticate(authenticator.AuthenticatePayload{
```
Comment (BenPope) [inline README.md]: Wow.  Can that be:
```go
app.Authenticate(auth.Payload{
```

File: authenticator/authenticator.go [hunk 4]
```diff
@@ -0,0 +1,129 @@
+package authenticator
```
Comment (BenPope) [inline authenticator/authenticator.go]: There are aspects of AuthN and AuthZ, should the package be `auth`?

File: authenticator/authenticator.go [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+const (
+	defaultTokenExpiry         = 24 * time.Hour
+	clientCredentialsGrantType = "client_credentials"
+	tokenType                  = "bearer"
+)
+
+// Authenticator specifies the public facing interface
+// for performing authentication and token generation
+type Authenticator interface {
+	Authenticate(
```
Comment (BenPope) [inline authenticator/authenticator.go]: Perhaps `Do`, like `http.Do`?

File: authenticator/types.go [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+type TokenResponse struct {
+	AccessToken string  `json:"access_token"`
+	TokenType   string  `json:"token_type"`
+	ExpiresIn   float64 `json:"expires_in"`
+}
+
+// AuthenticationResponse represents data that is returned
+// when making a call to the Authenticate method
+// It returns the status of the response, headers and the response body
+type AuthenticationResponse struct {
```
Comment (BenPope) [inline authenticator/types.go]: `Response`

File: authenticator/types.go [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+	Status  int         `json:"status"`
+	Headers http.Header `json:"headers"`
+	Body    interface{} `json:"body,omitempty"`
+}
+
+// AuthenticateOptions contains information to configure Authenticate method calls
+type AuthenticateOptions struct {
```
Comment (BenPope) [inline authenticator/types.go]: `Options`

File: authenticator/types.go [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+	UserID        *string
+	ServiceClaims map[string]interface{}
+	Su            bool
+	TokenExpiry   *time.Duration
+}
+
+// AuthenticatePayload specifies the grant type for the token
+type AuthenticatePayload struct {
```
Comment (BenPope) [inline authenticator/types.go]: `Payload`

File: authenticator/types.go [hunk 9]
```diff
@@ -35,9 +35,27 @@ type TokenResponse struct {
...
+	if a.Status != http.StatusOK {
+		return a.body.(*ErrorBody)
+	}
+
+	return nil
+}
+
+// TokenResponse returns the token returned by the response
+func (a *AuthenticationResponse) TokenResponse() *TokenResponse {
+	if a.Status != http.StatusOK {
```
Comment (BenPope) [inline authenticator/types.go]: Copy-paste error?

File: authenticator/types.go [hunk 10]
```diff
@@ -35,9 +35,27 @@ type TokenResponse struct {
...
+		return nil
+	}
+
+	return a.body.(*TokenResponse)
```
Comment (BenPope) [inline authenticator/types.go]: Should we prefer the panic, which makes it obvious if we've messed up, or do the type assert with the `ok bool`, and then return nil if not ok.  Might make it harder to determine what went wrong, but nicer for the user if it doesn't just panic.

Comment (vivangkumar) [reply authenticator/types.go]: Don't think we should panic. I'll go for the latter option.

File: authenticator/types.go (hunk 9 above)
Comment (vivangkumar) [reply authenticator/types.go]: Why? the return value is `nil` if we have a non success status code.

File: authenticator/authenticator.go (hunk 5 above)
Comment (vivangkumar) [reply authenticator/authenticator.go]: Want to keep it consistent with other platform SDK's..

Comment (BenPope) [reply authenticator/authenticator.go]: Sure, but there's quite a lot of stutter already. There's a trade-off between consistency and idiomatic for the language.

Comment (vivangkumar) [reply authenticator/authenticator.go]: Fair. `Do` is definitely more idiomatic.

File: auth/authenticator.go [hunk 11]
```diff
@@ -30,9 +30,9 @@ type authenticator struct {
 	keySecret  string
 }
 
-// New returns a new instance of an authenticator
+// NewAuthenticator returns a new instance of an authenticator
 // that conforms to the Authenticator interface
-func New(instanceID, keyID, keySecret string) Authenticator {
+func NewAuthenticator(instanceID, keyID, keySecret string) Authenticator {
```
Comment (BenPope) [inline auth/authenticator.go]: Just `New` was fine, now there's more stutter.  Most people don't distinguish AuthN and AuthZ, and those that do will figure it out.

File: authenticator/authenticator.go (hunk 5 above)
Comment (hamchapman) [reply authenticator/authenticator.go]: This is just an internal naming thing, right? There should definitely be an `Authenticate` function exposed in the outermost interface that customers interact with

File: authenticator/types.go (hunk 9 above)
Comment (BenPope) [reply authenticator/types.go]: I can't read.

File: authenticator/authenticator.go (hunk 5 above)
Comment (vivangkumar) [reply authenticator/authenticator.go]: Yes yes..

```go
instance.Authenticate(payload, options)
```

is the public interface.

File: instance/instance.go [hunk 12]
```diff
@@ -24,7 +24,7 @@ func init() {
 // Instance allows making HTTP requests to a service
 type Instance interface {
 	Request(ctx context.Context, options client.RequestOptions) (*http.Response, error)
-	Authenticator() auth.Authenticator
+	auth.Authenticator
```
Comment (BenPope) [inline instance/instance.go]: I don't think this is correct - it doesn't have the same interface; it wraps it, but with different naming.

File: auth/types.go [hunk 13]
```diff
@@ -5,13 +5,15 @@ import (
...
 
-// TokenWithExpiry represents a token that has an expiry time
+const GrantTypeClientCredentials = "client_credentials"
+
+// TokenWithExpiry represents a token that has an expiry time.
 type TokenWithExpiry struct {
-	Token     string
-	ExpiresIn float64
+	Token     string  // Token string
+	ExpiresIn float64 // Expiry in seconds
```
Comment (BenPope) [inline auth/types.go]: This should probably be:
```go
    ExpiresAt time.Time
```

File: auth/types.go [hunk 14]
```diff
@@ -23,24 +25,23 @@ func (e *ErrorBody) Error() string {
 	return e.ErrorDescription
 }
 
-// TokenResponse represents information that is returned on
-// generation of a token
+// TokenResponse represents information that is returned on generation of a token.
 type TokenResponse struct {
 	AccessToken string  `json:"access_token"`
 	TokenType   string  `json:"token_type"`
 	ExpiresIn   float64 `json:"expires_in"`
```
Comment (BenPope) [inline auth/types.go]: This should probably be:
```go
    ExpiresAt time.Time
```

File: auth/types.go [hunk 15]
```diff
@@ -68,15 +72,17 @@ func (a *Response) TokenResponse() *TokenResponse {
...
+	Su            bool                   // Indicates if token should contain the `su` claim
+	TokenExpiry   *time.Duration         // Optional token expiry (defaults to 24 hours)
 }
 
-// Payload specifies the grant type for the token
+// Payload specifies the grant type for the token.
+// Currently the only supported grant type is "client_credentials",
+// passing anything else other than this will return an error
 type Payload struct {
 	GrantType string
```
Comment (BenPope) [inline auth/types.go]: I don't really know what other options `GrantType` might be in the future, but is a string sufficiently expressive here?

File: client/client.go [hunk 16]
```diff
@@ -17,17 +17,17 @@ import (
 
 const authorizationHeader = "Authorization"
 
-// Client is a low level interface for clients of the elements protocol
+// Client is a low level interface for clients of the elements protocol.
 type Client interface {
 	Request(ctx context.Context, options RequestOptions) (*http.Response, error)
```
Comment (BenPope) [inline client/client.go]: This should conform to `http.RoundTripper`, but we can change that another time.

File: doc.go [hunk 17]
```diff
@@ -1,164 +1,16 @@
+// Package pusher provides an interface to interact with the Pusher platform.
```
Comment (BenPope) [inline doc.go]: Link to `pusher.com`?

File: doc_test.go [hunk 18]
```diff
@@ -0,0 +1,35 @@
...
+import (
+	"context"
+
+	"github.com/pusher/pusher-platform-go/client"
+	"github.com/pusher/pusher-platform-go/instance"
+)
+
+func Example() {
+	// Instantiate
+	app, err := instance.New(instance.Options{
```
Comment (BenPope) [inline doc_test.go]: We should prefer `instance` over `app`

Comment (vivangkumar) [reply doc_test.go]: Wanted to avoid collision with package name :(

**Pull request merged**